### ISSUE TITLE
Mypy unable to infer the correct function type

### ISSUE THREAD

**Issue by UniverseFly** [python/mypy#14430]
**Bug Report**

(A clear and concise description of what the bug is.)

**To Reproduce**

- https://gist.github.com/c93807dcc8512409d3e5c38a06ffa31c
- https://mypy-play.net/?mypy=latest&python=3.11&gist=c93807dcc8512409d3e5c38a06ffa31c

```python
from typing import TypeVar, Callable

T = TypeVar("T")
U = TypeVar("U")

def id(x: T) -> T:
    return x

def nothing(_: Callable[[U], int]) -> None:
    return None

nothing(id)
```

**Expected Behavior**

The type argument of `id` should be inferred as `int`

**Actual Behavior**

An error occurs complaining about a type mismatch between `Callable[[T], int]` and `Callable[[int], int]`.

**Your Environment**

As shown in the playground link.

**Additionals**

My original intention is to implement a simplified version of `functools.partial`, which is correctly type checked:

```python
def partial(func: Callable[Concatenate[T, P], U], arg: T) -> Callable[P, U]:
    def partially_called(*args: P.args, **kwargs: P.kwargs) -> U:
        return func(arg, *args, **kwargs)

    return partially_called
```

But when I actually call this `partial` on some generic function, the type checker starts to complain.

**Comment by thomkeh** [comment]
Similar to #13064, I think, but this new issue here has a shorter code example.

**Comment by hauntsaninja** [comment]
This code isn't correct. There is no guarantee that `id` returns an `int` given a `U`. In general, any time you see a function definition with a single type var (without bound), that indicates some misunderstanding. Pyright also rejects this code

**Comment by erictraut** [comment]
@UniverseFly, `U` has no valid meaning in your code sample because it appears only once in the type signature of `nothing`. The whole purpose behind a TypeVar within a function signature is to apply a constraint on that type. If a type variable appears only once (and it does not contain any upper bound or constraints of its own), then it serves to provide no such constraints. You might as well simply use `object` rather than `U`. And once you replace `U` with `object`, it should be clear why mypy is correct in generating a type violation error for `nothing(id)`.

**Comment by UniverseFly** [comment]
> @UniverseFly, `U` has no valid meaning in your code sample because it appears only once in the type signature of `nothing`. The whole purpose behind a TypeVar within a function signature is to apply a constraint on that type. If a type variable appears only once (and it does not contain any upper bound or constraints of its own), then it serves to provide no such constraints. You might as well simply use `object` rather than `U`. And once you replace `U` with `object`, it should be clear why mypy is correct in generating a type violation error for `nothing(id)`.

It makes sense that an unbound type parameter works like `object`, but here I have a more complex example, which I originally intended to use in my project and resulted in my creating this Issue:

- Gist URL: https://gist.github.com/ea10ba6bfa99a7c21d076517e4aa63e7
- Playground URL: https://mypy-play.net/?mypy=latest&python=3.11&gist=ea10ba6bfa99a7c21d076517e4aa63e7

```python
from typing import *

U = TypeVar("U")
T = TypeVar("T")
P = ParamSpec("P")

def partial(func: Callable[Concatenate[T, P], U], arg: T) -> Callable[P, U]:
    def partially_called(*args: P.args, **kwargs: P.kwargs) -> U:
        return func(arg, *args, **kwargs)

    return partially_called

def select(true: T, false: T, cond: bool, /) -> T:
    return true if cond else false

select_one = partial(select, 1)
```

Here I want to create a typed and simplified version of `functools.partial`. Every `TypeVar` is bound in this case. However, Mypy still complains a type mismatch. The expected behavior should be that the `select_one` definition can type-check.

**Comment by erictraut** [comment]
Yes, that appears to be a bug in mypy. It type checks without error in pyright.

@hauntsaninja, unless this is a known duplicate, you may want to re-open this issue to track this.
